src/clr-angular/forms/combobox/combobox.ts

```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
-  @ViewChild('input') input: ElementRef;
-  @ContentChild(ClrOptions) options: ClrOptions;
-  private subscription: Subscription;
+export class ClrCombobox<T> extends WrappedFormControl<ClrComboboxContainer>
+  implements ControlValueAccessor, LoadingListener {
+  @ViewChild('comboboxContainer', { static: true })
+  private comboboxContainer: ElementRef;
+  @ViewChild('textboxInput', { static: false })
+  textbox: ElementRef;
+  @ViewChild('trigger', { static: true })
+  trigger: ElementRef;
+  @ContentChild(ClrOptionSelected, { static: false })
+  optionSelected: ClrOptionSelected<T>;
+
+  public showCurrentSingleSelection: boolean = true;
+
+  private onChangeCallback;
+
+  private lastInputString: string = '';
+
+  protected index: number = 1;
 
   constructor(
+    vcr: ViewContainerRef,
+    injector: Injector,
+    @Self()
+    @Optional()
+    public control: NgControl,
+    protected renderer: Renderer2,
+    protected el: ElementRef,
+    public optionSelectionService: OptionSelectionService<T>,
+    public cdr: ChangeDetectorRef,
+    public commonStrings: ClrCommonStringsService,
     private toggleService: ClrPopoverToggleService,
-    private optionSelectionService: OptionSelectionService<T>,
-    private renderer: Renderer2,
-    private domAdapter: ComboboxDomAdapter
+    private positionService: ClrPopoverPositionService,
+    @Optional() private offsetService: ComboboxOffsetService,
+    private domAdapter: ComboboxDomAdapter,
+    private ariaService: AriaService,
+    private focusHandler: ComboboxFocusHandler
   ) {
-    // TODO: Might need to delay this to later in the lifecycle and use a BehaviorSubject instead
+    super(vcr, ClrComboboxContainer, injector, control, renderer, el);
+    if (control) {
+      control.valueAccessor = this;
+    }
+    // default to SingleSelectComboboxModel, in case the optional input [ClrMulti] isn't used
+    this.optionSelectionService.selectionModel = new SingleSelectComboboxModel<T>();
+
     this.initializeSubscriptions();
   }
 
-  private initializeSubscriptions(): void {
-    this.subscription = this.optionSelectionService.renderSelectionChanged.subscribe((option: ClrOption<T>) => {
-      this.renderSelection(option);
-    });
+  get openState(): boolean {
+    return this.toggleService.open;
   }
 
-  private renderSelection(selectedOption: ClrOption<T>): void {
-    if (this.input && selectedOption) {
-      this.domAdapter.clearChildren(this.input.nativeElement);
-      const clone: HTMLElement = this.domAdapter.cloneNode(selectedOption.elRef.nativeElement);
-      this.renderer.setAttribute(clone, 'contenteditable', 'false');
-      this.renderer.appendChild(this.input.nativeElement, clone);
+  public smartPosition: ClrPopoverPosition = {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.START,
+    content: ClrAlignment.START,
+  };
+
+  loadingStateChange(state: ClrLoadingState): void {
```

Comment:
I'm not seeing this called anywhere. Am I missing something?

Reply:
Combobox implements LoadingListener. We use this in the async case where we are waiting for options to load after opening the menu.